gtk4-macros/src/composite_template_callbacks_attribute.rs

```diff
@@ -0,0 +1,147 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use crate::attribute_parser::parse_attribute;
+use crate::util::*;
+use proc_macro2::{Ident, Span, TokenStream};
+use proc_macro_error::{abort, abort_call_site};
+use quote::quote;
+use syn::spanned::Spanned;
+
+pub const WRONG_PLACE_MSG: &str =
+    "This macro should be used on the `impl` block for a CompositeTemplate widget";
+
+pub fn impl_composite_template_callbacks(mut input: syn::ItemImpl) -> TokenStream {
+    let syn::ItemImpl {
+        attrs,
+        generics,
+        trait_,
+        self_ty,
+        items,
+        ..
+    } = &mut input;
+    if trait_.is_some() {
+        abort_call_site!(WRONG_PLACE_MSG);
+    }
+    let crate_ident = crate_ident_new();
+
+    let mut callbacks = vec![];
+    for item in items.iter_mut() {
+        if let syn::ImplItem::Method(method) = item {
+            let mut i = 0;
+            let mut attr = None;
+            while i < method.attrs.len() {
+                if method.attrs[i].path.is_ident("template_callback") {
```

Comment:
what would the template_callback provides? can't we just assume every method on the trait is a template callback?

Reply:
See @sdroege's comment a bit below :)

Reply:
> can't we just assume every method on the trait is a template callback?

I don't think that's safe - if you did that, you'd need a separate `impl` for the callbacks and non-callbacks, right?

Reply:
I don't think anyone has an Impl of the private trait though